### PR TITLE
Crown damage module

### PR DESCRIPTION
**Pull request by JessicaNeedham**
### Description:
This set of changes introduces a crown damage module into FATES. This reduces the crown area and biomass of tissues in the crown -leaves, sapwood, structure, storage and reproductive tissues. Damage can affect canopy or understory cohorts (or both) using the switches hlm_use_canopy_damage and hlm_use_understory_damage. Cohorts are split into damage classes, the number of which is set via the parameter ncrowndamage. The parameter damage_frac sets the fraction of each cohort that gets damaged in each damage event (currently hard coded to be once per year - but working on making this more flexible). The fates_damage_recovery_scalar parameter sets the degree of recovery i.e. allocation to regrow the crown versus staying on damaged allometric trajectory. Two optional parameters can be used to introduce damage-dependent mortality, which captures mortality from damage via mechanisms not represented in FATES. With default parameterisation this would be off. 

### Collaborators:
@ckoven @glemieux @rgknox  @rosiealice 

### Expectation of Answer Changes:
With the hlm_use_canopy_damage and hlm_use_understory_damage off runs should be bit for bit with master. Currently that is not the case and I'm looking into why. 
With these switches on then answers will not be bit for bit with master. There will be more cohorts as damage and recovery involve splitting cohorts into damage classes. Damaged cohorts will have smaller canopies which affects AGB and canopy dynamics. Depending on parameterisation carbon starvation mortality may increase. 

### Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided

### Test Results:

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

### CODE REVIEW

**Review by ckoven — COMMENTED** [review]
@JessicaNeedham this is really great, thanks!  I have some comments/questions, mostly just to improve clarity.

File: main/FatesHistoryInterfaceMod.F90 [hunk 1]
```diff
@@ -5934,6 +6613,203 @@ subroutine define_history_vars(this, initialize_variables)
           upfreq=2, ivar=ivar, initialize=initialize_variables, index = ih_resp_m_understory_si_scls )
 
 
+    ! CROWN DAMAGE VARIABLES
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply main/FatesHistoryInterfaceMod.F90:6616]: I'm deleting some of the history variables that aren't necessary. Some variables are also output along multiple combinations of dimensions. I could output each variable by just size x pft x damage class and let people aggregate in post processing?

File: main/EDTypesMod.F90 [hunk 2]
```diff
@@ -28,21 +28,21 @@ module EDTypesMod
   integer, parameter, public :: maxPatchesPerSite  = 14   ! maximum number of patches to live on a site
   integer, parameter, public :: maxPatchesPerSite_by_disttype(n_anthro_disturbance_categories)  = &
                                                      (/ 10, 4 /)  !!! MUST SUM TO maxPatchesPerSite !!!
-  integer,  public :: maxCohortsPerPatch = 100            ! maximum number of cohorts per patch
+  integer,  public :: maxCohortsPerPatch = 300            ! maximum number of cohorts per patch
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply main/EDTypesMod.F90]: Agreed, I've updated this.

File: main/FatesHistoryInterfaceMod.F90 [hunk 3]
```diff
@@ -2607,20 +2955,66 @@ subroutine update_history_dyn(this,nc,nsites,sites)
...
+                            ccohort%c_area  + &
+                            (ccohort%lmort_direct + ccohort%lmort_collateral + ccohort%lmort_infra) * &
+                            ccohort%c_area * sec_per_day * days_per_year 
+
+                      
+                       hio_m3_mortality_canopy_si_scls(io_si,scls) = hio_m3_mortality_canopy_si_scls(io_si,scls)+&
+                            ccohort%cmort * ccohort%n
+                        
+                       ! damage variables - canopy
+                       if(hlm_use_canopy_damage .eq. itrue .or. hlm_use_understory_damage .eq. itrue) then
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply main/FatesHistoryInterfaceMod.F90:2859]: This block needs to be inside the 'if canopy' block.

File: parameter_files/fates_params_default.cdl [hunk 4]
```diff
@@ -153,7 +153,22 @@ variables:
 	double fates_branch_turnover(fates_pft) ;
 		fates_branch_turnover:units = "yr" ;
 		fates_branch_turnover:long_name = "turnover time of branches" ;
-	double fates_c2b(fates_pft) ;
+        double fates_damage_frac(fates_pft) ;
+                fates_damage_frac:units = "fraction";         
+                fates_damage_frac:long_name = "fraction of cohort damaged in each damage event";
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply parameter_files/fates_params_default.cdl:158]: The timing of damage events is specified in DamageMainMod in the is_it_damage_time subroutine (modelled off the is_it_logging_time subroutine). At the moment there is one damage event per year on Jan 1st but users should be able to set specific dates or frequencies. I've updated this comment.

File: main/FatesHistoryInterfaceMod.F90 [hunk 5]
```diff
@@ -5194,6 +5830,16 @@ subroutine define_history_vars(this, initialize_variables)
           long='flux of biomass carbon from live to dead pools from mortality of understory plants',use_default='active',&
           avgflag='A', vtype=site_r8, hlms='CLM:ALM', flushval=0.0_r8,    &
           upfreq=1, ivar=ivar, initialize=initialize_variables, index = ih_understory_mortality_carbonflux_si )
+    
+    call this%set_history_var(vname='MORTALITY_CROWNAREA_CANOPY', units = 'm2/ha/year',               &
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply main/FatesHistoryInterfaceMod.F90]: Updated names and units.

File: main/EDTypesMod.F90 [hunk 6]
```diff
@@ -340,6 +342,9 @@ module EDTypesMod
                                                          ! (below ground)
      real(r8) ::  froot_mr                               ! Live fine root   maintenance respiration: kgC/indiv/s
 
+     !DAMAGE
+     real(r8) :: branch_frac                             ! Fraction of aboveground biomass in branches
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply main/EDTypesMod.F90]: I've simplified the code so that branch_frac is no longer a cohort variable. Also updated comments where it is declared.

**Review by glemieux — CHANGES_REQUESTED** [review]
Initial review due to regression test failure in debug mode.  One other minor suggestion as well.

File: main/FatesParameterDerivedMod.F90 [hunk 7]
```diff
@@ -85,11 +118,78 @@ subroutine Init(this,numpft)
...
+          if(i < nlevdamage) then
+             ! fraction damaged get split according to class width
+             this%damage_transitions(i,i+1:nlevdamage,ft) = damage_frac * &
+                  class_widths(i+1:nlevdamage)/ SUM(class_widths(i+1:nlevdamage))  
+          end if
+          ! Make sure it sums to one - they have to go somewhere
+          this%damage_transitions(i, :, ft) = this%damage_transitions(i, :, ft)/SUM(this%damage_transitions(i, :, ft))
+       end do
+
+        write(fates_log(),'(a/,5(F12.6,1x))') 'annual transition matrix : ', this%damage_transitions(:,:,ft)
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply main/FatesParameterDerivedMod.F90:186]: Oh, good catch, I agree with putting it in a debug  statement. Or we can delete it. It is supposed to write out the damage transition matrix but I couldn't get  it to work with a variable number of damage bins.

File: biogeophys/FatesPlantRespPhotosynthMod.F90 [hunk 8]
```diff
@@ -640,17 +658,37 @@ subroutine FatesPlantRespPhotosynthDrive (nsites, sites,bc_in,bc_out,dtime)
...
+
+                         sapw_c_undamaged = sapw_c / &
+                              (1.0_r8 - (agb_frac * branch_frac * crown_reduction))
+                         ! Undamaged below ground portion
+                         sapw_c_bgw = sapw_c_undamaged * (1.0_r8 - agb_frac)
+
+                         ! Damaged aboveground portion
+                         sapw_c_agw = sapw_c - sapw_c_bgw
+                         
+                      end if
```
[the remark this answers is not in the document]
Comment (JessicaNeedham) [reply biogeophys/FatesPlantRespPhotosynthMod.F90:681]: Looks good to me.

**Comment by glemieux** [comment]
Regression testing on Cheyenne is complete.  All expected tests pass b4b.  Note that the tree damage test mod which wasn't generating baselines prior to this PR is now running and generating baselines successfully.  Also, the addition of `MORTALITY_CROWNAREA_CANOPY` and `MORTALITY_CROWNAREA_UNDERSTORY` is resulting in expected `FIELDLIST` DIFFs for those test mods that includes all active history output.

Folder location: `/glade/u/home/glemieux/scratch/ctsm-tests/tests_pr788-crowndamage-2`

**Review by glemieux — APPROVED** [review]

**Pull request merged**